### PR TITLE
Switch to automatic dependency injection of Utility APIs

### LINKED ISSUE #2182: Use automatic dependency injection for plugin-provided Utility APIs
## Feature Suggestion

The [list of APIs in the app](https://github.com/spotify/backstage/blob/6d752757c2e2d6f9e6a2892cb5344e1b614a595d/packages/app/src/apis.ts#L86) is becoming pretty large, somewhat defeating the purpose of the clear wiring of the APIs. We initially choose a manual wiring of the APIs to avoid magic and making it more transparent how plugins are added to the app. We also knew that we might have to revisit that choice in case the number of APIs grew rapidly, which is currently the case.

I'm suggesting we switch to an automated DI solution where the plugins provide factories for their APIs, similar to what's being done in dev apps [here](https://github.com/spotify/backstage/blob/6d752757c2e2d6f9e6a2892cb5344e1b614a595d/packages/dev-utils/src/devApp/apiFactories.ts#L44). The factories would still be possible to override in the app, either with a new factory for the same `ApiRef`, or with a concrete API implementation.

## Possible Implementation

Promote the `ApiTestRegistry` to be the main API registry implementation (breaking change), and add something like a `factories` field to `createPlugin`. The initial change would be to just enable the feature, and we can then start migrating APIs to factories in separate changes.

## Context

This is part of streamlining the addition of plugins to the app. It's also part of making sure that plugins own the default instantiation of their API, and that we way of reading config is encoded into the plugin and not the app.

### PR DESCRIPTION
**Pull request by Rugvip**
Fixes #2182

Will add doc changes to this PR once we're happy with the changes.

Previous design was a single API registry where all APIs where wired up in `apis.ts`, simple but tedious and a lot of code required in the App.

This switches out to a more complex setup, but where the work of wiring up plugins in moved out from the app, making it significantly easier to add plugins to an existing app. `apis.ts` still exists, but it's only used for overriding the default api factories shipped in `@backstage/core` and each plugin.

Instead of directly instantiating all APIs and startup, we now lazy-load them as they are asked for. There's a quick sanity check at startup to make sure that there are no cycles in the dependency tree, but other than that failures will be encountered later. The instantiation is still synchronous, which is very much intentional, as I don't think we want plugins to do any kind of work at startup.

Couple of naming things to discuss:

- I opted for referring to the ApiFactories as just `api` in the external API. Since there's no other way to instantiate an API than using a factory, I felt that `apiFactories` and `defaultApiFactories` etc. were a bit redundant and went with the shorter form.
- Each ApiFactory earlier had an `implements` field, which I renamed to `api`. I think `api` captures the intention a bit better and is a bit more readable. I'm trying to skip having to think about "creating API factories", and make the mental model be that we're creating APIs. Another option here would be `apiRef`, since that's the actual value, but I feel that `api` is nicer, and typescript should take care of avoiding any confusion there anyway. It also aligns with `useApi`, where you're also actually passing in an `ApiRef`.
- Related to the above, the function to create ApiFactories is called `createApiFactory`. Here I did feel like `createApi` would be a bit too misleading, and I left in the `Factory` word just to give a hint at what it is we're actually doing.

### CODE REVIEW

File: plugins/catalog/src/plugin.ts [hunk 1]
```diff
@@ -14,8 +14,21 @@
...
+import { catalogApiRef } from './api/types';
+import { CatalogClient } from './api/CatalogClient';
 
 export const plugin = createPlugin({
   id: 'catalog',
+  apis: [
+    createApiFactory({
+      api: catalogApiRef,
+      deps: { discoveryApi: discoveryApiRef },
+      factory: ({ discoveryApi }) => new CatalogClient({ discoveryApi }),
```
Comment (Rugvip) [inline plugins/catalog/src/plugin.ts:31]: Here's an example of what it looks like from a plugin's point of view

**Comment by Rugvip** [comment]
Not intending for this to build yet, will update docs and create-app etc once we've settled any discussions :grin:

Please do review still :>

**Review by marcuseide — APPROVED** [review]
👍

**Comment by dtuite** [comment]
Will this require changes in plugins?

**Comment by Rugvip** [comment]
@dtuite preferably, but not immediately. Plugins can still choose to not supply default factories for their APIs, but then they have to be wired up in the app like before.

e.g. https://github.com/spotify/backstage/blob/62c2971a4aa923c1a49982d5bf2cba0145c62525/packages/app/src/apis.ts#L70

For the Travis and GH PRs plugins to be updated we'll need a new release with these changes, so that they can start exposing the factories, and then remove them from the app.

**Comment by dtuite** [comment]
@Rugvip we will work on that once this is merged. We're also working on migrating to the new routing API and tabs on the catalog entity page.

**Review by hooloovooo — APPROVED** [review]

**Review by emmaindal — APPROVED** [review]

**Comment by Rugvip** [comment]
Adding docs in a followup PR so we can have a more focused review

**Comment by Rugvip** [comment]
#2334

**Comment by Rugvip** [comment]
@dtuite Yep, just making sure #2200 is all good before we release, i.e. waiting for that to be closed.

**Pull request merged**